src/Microsoft.AspNet.Mvc.ModelBinding/ModelBindingContext.cs

```diff
@@ -71,12 +70,14 @@ public object Model
             set
             {
                 IsModelSet = true;
-
+                ModelStateKey = ModelName;
```

Comment:
I don't understand this. 

It sounds like what you're saying is that the contract of `IModelBinder` needs to change to support what we're doing.  What if we did something like this instead?

```
public interface IModelBinder
{
    Task<ModelBindingResult> BindModelAsync(ModelBindingContext);
}

public class ModelBindingResult
{
    public object Model {}
    public string Key {}
}
```